### PR TITLE
Support host mapping for virtual hosting

### PR DESCRIPTION
**Pull request by technosophos**
This documents and implements virtual host mapping.

### CODE REVIEW

**Comment by technosophos** [comment]
Okay, so here is a problem case:

```toml
[[module]]
# Example executing a Web Assembly Text file
route = "/"
module = "examples/hello.wat"

[[module]]
# Example running a WASM file.
route = "/hello/..."
module = "examples/hello.wasm"
host = "example.com"
```

This command will match the second route:

```console
curl -v -H "Host: example.com" localhost:3000/hello/hello
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET /hello/hello HTTP/1.1
> Host: example.com
> User-Agent: curl/7.64.1
> Accept: */*
```

That is correct. But should this one match the first route, or be a 404?

```
curl -v -H "Host: example.com" localhost:3000/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.64.1
> Accept: */*
```

**Comment by bacongobbler** [comment]
Hmm... I think it should match the first route. If no virtual host is defined by the route, then the `Host` header should be ignored IMO. It's only when the operator opts into the vhost rule that it should be evaluated.

**Comment by radu-matei** [comment]
I think it should be a 404.
We don't do fallback routes for any other scenarios, and I can't think of a reason why it should be the case here.

**Comment by bacongobbler** [comment]
While I couldn't find any information on how HTTP servers should handle Host headers, there is a type of attack known as an HTTP Host Header attack (aptly named). If the server implicitly trusts the Host header and fails to validate or escape it properly, an attacker may be able to use this input to inject harmful payloads and manipulate server-side behaviour. A successful attack can be defined as dispatching requests to the first virtual host on the list.

In other words, we should not fallback. It should be a 404 request.

A good summary of Host header attacks can be found here: https://infosecwriteups.com/identifying-escalating-http-host-header-injection-attacks-7586d0ff2c67

**Comment by technosophos** [comment]
Just to be clear, what we are saying is:

If the `Host` header is set to something other than "the server default" (e.g. localhost) and a route matches the path, but not the host, then the route should return a 404.

Does that sound right?

(Then I just need to figure out what "the server default" is)

**Comment by technosophos** [comment]
PR is updated to require domain matches. I am still looking into what the default host should be. Right now, it's hard-coded to `localhost`, but I think I should check it against the system's hostname, too.

**Comment by radu-matei** [comment]
Probably allowing an environment variable to control the host might be reasonable?

**Comment by radu-matei** [comment]
Probably allowing an environment variable and a flag to control the host might be reasonable?

**Comment by technosophos** [comment]
So I got as far as implementing a function that checks if the hostname is one of `localhost`, `localhost.localdomain`, `127.0.0.1`, and whatever the the system hostname is.

But I am not sure if that is the right behavior. I'm wondering if maybe the default should be `localhost`, and a CLI flag/env var should override it (like @radu-matei suggested)... and we should leave it at that.

That would prevent some of the attacks @bacongobbler pointed to in that article. And I think using `localhost` is a reasonable default, since that matches a developer's usual environment.

**Comment by radu-matei** [comment]
`localhost` as the default that can be overridden by the config seems reasonable to me.

**Comment by technosophos** [comment]
Okay, this is ready for review now

File: src/lib.rs [hunk 1]
```diff
@@ -78,6 +86,17 @@ pub fn load_modules_toml(
 /// The configuration for all modules in a WAGI site
 #[derive(Clone, Deserialize)]
 pub struct ModuleConfig {
+    /// The default hostname to use if none is supplied.
+    ///
+    /// If this is not set, the default hostname is `localhost`.
+    ///
+    /// Incoming HTTP requests MUST match a hostn name, or else they will not be processed.
```
Comment (radu-matei) [inline src/lib.rs:93]: typo: `hostn`

**Review by radu-matei — APPROVED** [review]
LGTM, with one small typo.

**Comment by technosophos** [comment]
Typo fixed
